src/main/java/reactor/kafka/receiver/observation/KafkaRecordReceiverContext.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2018-2023 VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.kafka.receiver.observation;
+
+import java.nio.charset.StandardCharsets;
+
+import io.micrometer.observation.transport.ReceiverContext;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+
+/**
+ * {@link ReceiverContext} for {@link ConsumerRecord}s.
+ *
+ * @author Gary Russell
+ * @author Artem Bilan
+ * @since 1.4
+ *
+ */
+public class KafkaRecordReceiverContext extends ReceiverContext<ConsumerRecord<?, ?>> {
+
+    private final String clientId;
+
+    private final ConsumerRecord<?, ?> record;
+
+    public KafkaRecordReceiverContext(ConsumerRecord<?, ?> record, String clientId, String kafkaServers) {
+        super((carrier, key) -> {
+            Header header = carrier.headers().lastHeader(key);
+            if (header == null) {
+                return null;
+            }
+            return new String(header.value(), StandardCharsets.UTF_8);
+        });
+        setCarrier(record);
+        this.record = record;
+        this.clientId = clientId;
```

Comment:
For clientId and topic maybe we could leverage the context from inherited Observartion.Context ? It has a map with put/get methods. Later that context can be propagated to span tags/attributes, metrics labels. etc.

Reply:
I'm not sure in this question. See `DefaultKafkaReceiverObservationConvention`. That's exactly where we populate our own key-values. Yes, in the end everything ends up in the mentioned map, but how is that really convenient to deal with the `Map` API instead of our own getters for the mentioned convention?

Reply:
I've missed that part where it adds those keyvalues to observation context. In that case can we add to the observation keyvalues the topic and partition ? It is useful sometimes to know the exact source of the event. Something similar is currently done by sleuth implementation for it's kafka instrumentations.